### PR TITLE
AZP: analyzer for iodemo

### PR DESCRIPTION
**Pull request by avildema**
add analyzer for io_demo test

### CODE REVIEW

File: buildlib/io_demo/io-demo.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            uptime: ${{ test.Value.uptime }}
+      maxParallel: 1
+
+    variables:
+      workspace: $[ dependencies.io_build.outputs['build.workspace'] ]
+
+    displayName: "io_demo: "
+    steps:
+      - checkout: none
+
```
Comment (yosefe) [inline buildlib/io_demo/io-demo.yml:69]: remove space

File: buildlib/io_demo/iodemo_analyzer.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+re_error = re.compile(r".*(error|assert|backtrace|segmentation).*", re.I)
+re_warning = re.compile(r".*warn.*", re.I)
+
+
+def in_allow_list(line, is_allow_list):
+    is_allow_error = False
+    if is_allow_list:
+        s = re_allow_list.search(line)
+        if s:
+            is_allow_error = True
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:49]: return True

File: buildlib/io_demo/iodemo_analyzer.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                    data[d[0]] = {
+                        'pos': int(d[1]),
+                        'timestamp': ts,
+                        'timestamp_rx': rx_ts,
+                        'timestamp_tx': tx_ts,
+                        }
+        # Burn After Reading
+        open(seek_file, 'w').close()
+    return data
+
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:73]: another space line

File: buildlib/io_demo/iodemo_analyzer.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+def read_file(filename):
+    data = ""
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:75]: data = []

File: buildlib/io_demo/iodemo_analyzer.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    with open(filename) as f:
+        data = f.readlines()
+    return data
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:78]: can we reuse this function in process_seek?

File: buildlib/io_demo/iodemo_analyzer.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    client_list = []
+    server_list = []
+    for f in os.listdir(directory):
+        filename = os.path.join(directory, f)
+        if os.path.isfile(filename) and "_client_" in f:
+            client_list.append(filename)
+        if os.path.isfile(filename) and "_server_" in f:
+            server_list.append(filename)
+    return client_list, server_list
+
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:91]: another space line

File: buildlib/io_demo/iodemo_analyzer.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+            # print("analyze {}".format(log))
+            while True:
+                line = f.readline()
+
+                if not line:
+                    break
+
+                m = re_error.match(line, re.IGNORECASE) or re_warning.match(line, re.IGNORECASE)
+                if m and not in_allow_list(line, is_allow_list):
+                    #print("log {} [{}] contains error: {}".format(log, i, line))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:104]: remove

File: buildlib/io_demo/iodemo_analyzer.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+                    raise Exception("Contains error: {}\nLog {}:\nLine {}".format(line, log, line))
+
+
+def process_client(files, threshold, seek_file, is_allow_list):
+
+    seek_data = process_seek(seek_file)
+
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:111]: remove space lines

File: buildlib/io_demo/iodemo_analyzer.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    for log in files:
+        with open(log) as f:
+            # print("analyze {}".format(log))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:114]: remove

File: buildlib/io_demo/iodemo_analyzer.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+                line = f.readline()
+
+                if not line:
+                    if seek_file and cur_traffic_date:
+                        pos = f.tell()
+                        with open(seek_file, 'a+') as s:
+                            s.write("{}\t{}\t{}\t{}\t{}\n".format(log, pos, cur_traffic_date, zero_rx_ts, zero_tx_ts))
+                    break
+
+                t_r = re_timestamp.match(line)
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:134]: t_r  -> need a more meaningful name

File: buildlib/io_demo/iodemo_analyzer.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+                    cur_traffic_date = m.group(1)
+                    date_traffic = float(cur_traffic_date)
+                    curr_traffic_ts = datetime.datetime.fromtimestamp(date_traffic)
+                    rx = float(m.group(2))
+                    min_server_rx=int(m.group(3))
+
+                    if min_server_rx == 0 and zero_rx_ts:
+                        delta = curr_traffic_ts - datetime.datetime.fromtimestamp(zero_rx_ts)
+                        if delta.total_seconds() > threshold * 60:
+                            #print ("log {} [{}] have read min:0 servers {} minutes (more threshold:{})".format(log, i, delta.total_seconds()/60.0, threshold))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:158]: remove

File: buildlib/io_demo/iodemo_analyzer.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+                            raise Exception("Have read min:0 servers {} minutes (more threshold:{})\nlog {}:\nLine {}".format(delta.total_seconds()/60.0, threshold, log, line))
+                    else:
+                        zero_rx_ts = date_traffic
+
+                    if not rx:
+                        #print("log {} [{}] have rx or tx zero speed: {} ".format(log, i, line))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:164]: remove

File: buildlib/io_demo/iodemo_analyzer.py [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+                    date = float(t_r.group(1))
+                    curr_ts = datetime.datetime.fromtimestamp(date)
+                    if not prev_traffic_ts:
+                        prev_traffic_ts = curr_ts
+
+                i += 1
+                m_rx = re_traffic_read.match(line)
+                m_tx = re_traffic_write.match(line)
+
+                m = None
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:145]: m, m_rx, m_tx -> better names

File: buildlib/io_demo/iodemo_analyzer.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+                        raise Exception("Have read zero speed:\nLog {}:\nLine {}".format(log, line))
+                    prev_traffic_ts = curr_traffic_ts
+
+                if m_tx:
+                    m = m_tx
+                    cur_traffic_date = m.group(1)
+                    date_traffic = float(cur_traffic_date)
+                    curr_traffic_ts = datetime.datetime.fromtimestamp(date_traffic)
+                    tx = float(m.group(2))
+                    min_server_tx=int(m.group(3))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:174]: space around =

File: buildlib/io_demo/iodemo_analyzer.py [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+
+                    if min_server_tx == 0 and zero_tx_ts:
+                        delta = curr_traffic_ts - datetime.datetime.fromtimestamp(zero_tx_ts)
+                        if delta.total_seconds() > threshold * 60:
+                            #print("log {} [{}] have write min:0 servers {} minutes (more threshold:{})".format(log, i, delta.total_seconds()/60.0, threshold))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:179]: remove

File: buildlib/io_demo/iodemo_analyzer.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+                            raise Exception("Have write min:0 servers {} minutes (more threshold:{})\nLog {}:\nLine {}".format(delta.total_seconds()/60.0, threshold, log, line))
+                    else:
+                        zero_tx_ts = date_traffic
+
+                    if not tx:
+                        #print("log {} [{}] have rx or tx zero speed: {} ".format(log, i, line))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:185]: remove

File: buildlib/io_demo/iodemo_analyzer.py [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+                        raise Exception("Have write zero speed:\nLog {}:\nLine {}".format(log, line))
+
+                    prev_traffic_ts = curr_traffic_ts
+
+
+                if m and prev_traffic_ts:
+                    delta = curr_traffic_ts - prev_traffic_ts
+                    if delta.total_seconds() > threshold * 60:
+                        #print("log {} [{}] have delta {} more {} minutes".format(log, i, delta.total_seconds()/60.0, threshold))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:194]: remove

File: buildlib/io_demo/iodemo_analyzer.py [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+                        raise Exception("Have delta {} more {} minutes\nLog {}:\nLine {}".format(delta.total_seconds()/60.0, threshold, log, line))
+
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:196]: remove one space line

File: buildlib/io_demo/iodemo_analyzer.py [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+
+                if not m:
+                    m = re_error.match(line, re.IGNORECASE)
+                    if m:
+                        if not in_allow_list(line, is_allow_list):
+                            #print("log {} [{}] contains error: {}".format(log, i, line))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:202]: remove

File: buildlib/io_demo/iodemo_analyzer.py [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+            is_allow_error = True
+    return is_allow_error
+
+
+def process_seek(seek_file):
+    data = {}
+    if seek_file and os.path.exists(seek_file):
+        with open(seek_file) as f:
+            for line in f.readlines():
+                if line:
```
[the remark this answers is not in the document]
Comment (avildema) [reply buildlib/io_demo/iodemo_analyzer.py:58]: it is Ok, not need additional condition

File: buildlib/io_demo/iodemo_analyzer.py (hunk 5 above)
Comment (avildema) [reply buildlib/io_demo/iodemo_analyzer.py:78]: it is unused function - removed

**Comment by avildema** [comment]
bot:pipe:retest

**Comment by avildema** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
@avildema pls avoid force-push next time
see https://github.com/openucx/ucx/blob/master/.github/CONTRIBUTING.md

File: buildlib/io_demo/az-stage-io-demo.yaml [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+        $(workspace)/../test/apps/iodemo/io_demo \
+            -d 512:524288 \
+            -o read,write \
+            -i 0 \
+            -w 16 \
+            -t 10 \
+            ${{ parameters.iodemo_args }}
+
+  displayName: Launch with run_io_demo.sh ( ${{ parameters.name }} )
+  timeoutInMinutes: 2
```
Comment (yosefe) [inline buildlib/io_demo/az-stage-io-demo.yaml:50]: let's increase to 10min

File: buildlib/io_demo/iodemo_analyzer.py [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+            server_list.append(filename)
+    return client_list, server_list
+
+
+def process_server(files, is_allow_list):
+    for log in files:
+        with open(log) as f:
+            while True:
+                line = f.readline()
+
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:93]: remove space

File: buildlib/io_demo/iodemo_analyzer.py [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+            cur_traffic_date = ""
+            prev_traffic_ts = seek_data.get(log, {}).get('timestamp', 0)
+            zero_rx_ts = seek_data.get(log, {}).get('timestamp_rx', 0)
+            zero_tx_ts = seek_data.get(log, {}).get('timestamp_tx', 0)
+            pos_prev = seek_data.get(log, {}).get('pos', 0)
+            f.seek(pos_prev)
+            i = 0
+            while True:
+                line = f.readline()
+
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:117]: remove space

File: buildlib/io_demo/iodemo_analyzer.py [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+                if not line:
+                    if seek_file and cur_traffic_date:
+                        pos = f.tell()
+                        with open(seek_file, 'a+') as s:
+                            s.write("{}\t{}\t{}\t{}\t{}\n".format(log, pos, cur_traffic_date, zero_rx_ts, zero_tx_ts))
+                    break
+
+                m_ts = re_timestamp.match(line)
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:125]: m_ts -> ts_match

File: buildlib/io_demo/iodemo_analyzer.py [hunk 25]
```diff
@@ -0,0 +1,232 @@
...
+                if m_ts:
+                    date = float(m_ts.group(1))
+                    curr_ts = datetime.datetime.fromtimestamp(date)
+                    if not prev_traffic_ts:
+                        prev_traffic_ts = curr_ts
+
+                i += 1
+                m_rx = re_traffic_read.match(line)
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:133]: read_match
write_match

File: buildlib/io_demo/iodemo_analyzer.py (hunk 13 above)
Comment (yosefe) [reply buildlib/io_demo/iodemo_analyzer.py:136]: m -> current_match

File: buildlib/io_demo/iodemo_analyzer.py [hunk 26]
```diff
@@ -0,0 +1,232 @@
...
+                    cur_traffic_date = m.group(1)
+                    date_traffic = float(cur_traffic_date)
+                    curr_traffic_ts = datetime.datetime.fromtimestamp(date_traffic)
+                    rx = float(m.group(2))
+                    min_server_rx = int(m.group(3))
+
+                    if min_server_rx == 0 and zero_rx_ts:
+                        delta = curr_traffic_ts - datetime.datetime.fromtimestamp(zero_rx_ts)
+                        if delta.total_seconds() > threshold * 60:
+                            raise Exception("Have read min:0 servers {} minutes (more threshold:{})\nlog {}:\nLine {}".format(delta.total_seconds()/60.0, threshold, log, line))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:149]: split long line with `\`

File: buildlib/io_demo/iodemo_analyzer.py [hunk 27]
```diff
@@ -0,0 +1,232 @@
...
+                    cur_traffic_date = m.group(1)
+                    date_traffic = float(cur_traffic_date)
+                    curr_traffic_ts = datetime.datetime.fromtimestamp(date_traffic)
+                    tx = float(m.group(2))
+                    min_server_tx=int(m.group(3))
+
+                    if min_server_tx == 0 and zero_tx_ts:
+                        delta = curr_traffic_ts - datetime.datetime.fromtimestamp(zero_tx_ts)
+                        if delta.total_seconds() > threshold * 60:
+                            raise Exception("Have write min:0 servers {} minutes (more threshold:{})\nLog {}:\nLine {}".format(delta.total_seconds()/60.0, threshold, log, line))
```
Comment (yosefe) [inline buildlib/io_demo/iodemo_analyzer.py:168]: split long line with `\`
same for other very long lines

File: buildlib/io_demo/az-stage-io-demo.yaml (hunk 21 above)
Comment (avildema) [reply buildlib/io_demo/az-stage-io-demo.yaml:50]: @yosefe 
it's timeout for azure. do I increase time for run_io_demo

Comment (yosefe) [reply buildlib/io_demo/az-stage-io-demo.yaml]: for run_io_demo.sh:
```
	echo "  -d|--duration <seconds>     How much time to run the application"$(show_default_value duration)
```

**Review by yosefe — APPROVED** [review]
pls squash

**Comment by yosefe** [comment]
error is relevant:
https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=11545&view=logs&jobId=33b344a0-dd2c-560b-13f8-88b3d0888043&j=33b344a0-dd2c-560b-13f8-88b3d0888043&t=9d92d56b-8d6c-5915-64c2-2d02b4e9e40f
```
python: can't open file 'buildlib/io_demo/iodemo_analyzer.py': [Errno 2] No such file or directory
```

**Review by yosefe — APPROVED** [review]

**Comment by avildema** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
@avildema there is a conflict now

**Review by yosefe — APPROVED** [review]
pls squash

**Review by yosefe — APPROVED** [review]

File: buildlib/io_demo/io-demo.yml [hunk 28]
```diff
@@ -33,9 +33,10 @@ jobs:
       - bash: |
           set -eEx
           ./autogen.sh
-          ./contrib/configure-release --prefix=$PWD/__install
+          ./contrib/configure-release --prefix==$(Build.Repository.LocalPath)/install
```
Comment (yosefe) [inline buildlib/io_demo/io-demo.yml:36]: duplicate =

Comment (avildema) [reply buildlib/io_demo/io-demo.yml]: done

**Review by yosefe — APPROVED** [review]
pls squash

**Pull request merged**